tgui/packages/tgui-panel/chat/middleware.js

```diff
@@ -85,9 +85,9 @@ export const chatMiddleware = (store) => {
       loadChatFromStorage(store);
     }
     if (type === 'chat/message') {
-      // Normalize the payload
-      const batch = Array.isArray(payload) ? payload : [payload];
-      chatRenderer.processBatch(batch);
+      const payload_obj = JSON.parse(payload);
+      Byond.sendMessage('chat/acknowledge', payload_obj.sequence);
```

Comment:
Sending acks on each message might break Byond's AFK detector, because it treats incoming topic calls as activity.

I don't know what to suggest though. I know there is a ping system in tgchat that stops sending pings if user is AFK, but outgoing message acks will reset the AFK counter on the Byond's side, so you can't rely on it to tell whether the user is AFK.

Reply:
Not really sure what else we could do here. We need the ability for a client to acknowledge that a payload was received

Reply:
Well, need is a strong word; but it's the only way I can think of to add a reliability layer

Reply:
you can do an incrementing counter that indicates wher ethe server and client both think the number of chat messages is at.

if you want to add replay on top of that to make it reliable you'd need to record the last n messages sent to each client, which is going to cost some memory, but once the sequence number gets out of sync the client can ask the server to send n many back messages that it's missing and redisplay them.

Reply:
Sounds like a real solution, replay messages only if client detects it is out of sync.

Reply:
done